README.md
# hsimpy




<p align="center">
    <img src ="./doc/img/logo.png" />
</p>


> **web page:** https://howsecureismypassword.net/

> **client-side:** https://github.com/howsecureismypassword/hsimp



## ABOUT hsimpy

> How Secure Is My Password?

Running on Python3.
Rather than just saying a password is "weak" or "strong", *How Secure is My Password?* lets your users know how long it would take someone to crack their password.




## EASY TO USE

### INSTALLING

hsimpy is a python package to be used called by a python module. It is needed to install the package to the usage.
In you *requirements.txt* file. Should be declared.

In your project where hsimpy is a dependencia, enter the command.

```bash
pip3 install -r requirements.txt
```


### USAGE

``` python
from hsimpy import Hsimpy

PWD = Hsimpy('Passw0rd!')

print('possible_characters: ' + PWD.possible_characters)
print('possible_combinations: ' + PWD.possible_combinations)
print('time_in_seconds: ' + PWD.time_in_seconds)
print('security_level: ' + PWD.security_level)
```


#### hsmipy.Hsimpy module configuration
hsmip module has a small config. We are using:

``` python
 {
    "calculationsPerSecond": 1e10, # 10 billion
    "good": 31557600e6, # 1 million years
    "ok": 31557600 # 1 year
}
```

Today, the data used is:

``` python
self.conf = {
    "calculations_per_second": 3900000,  # 1e10 10 billion
    "good": 31557600e3,  # 31557600e6 1 million years
    "ok": 31557600e1  # 31557600 1 year
}

```

#### character.Character module configuration
Sharacter module has a small config. We are using:

``` python
 {
    "ASCII Control Character": ["[\\u0000-\\u001F]", 32],
    "ASCII Lowercase": ["[a-z]", 26],
    "ASCII Uppercase": ["[A-Z]", 26],
    "ASCII Numbers": ["\\d", 10],
    "ASCII Top Row Symbols": ["[-!@£#$%^&*()=+_]", 15],
    "ASCII Other Symbols": ["[\\s\\?\\/\\.>,<`~\\|;:\\]}\\[{'\"\\\\]", 19],
    "Unicode Latin 1 Supplement": ["[\\u00A1-\\u00A2\\u00A4-\\u00FF]", 93],
    "Unicode Latin 1 Supplement Non Standard": ["[\\u0080-\\u00A0]", 33],
    "Unicode Latin Extended A": ["[\\u0100-\\u017F]", 128],
    "Unicode Latin Extended B": ["[\\u0180-\\u024F]", 208],
    "Unicode Latin Extended C": ["[\\u2C60-\\u2C7F]", 32],
    "Unicode Latin Extended D": ["[\\uA720-\\uA7FF]", 29],
    "Unicode Cyrillic Uppercase": ["[\\u0410-\\u042F]", 32],
    "Unicode Cyrillic Lowercase": ["[\\u0430-\\u044F]", 32]
}
```




## DEVELOPMENT

When not exist a pylintrc file. We can create using command below:

``` bash
pylint --generate-rcfile > ~/.

```




### USING PYLINT AND PYCODESTYLE

``` bash
pylint ./hsimpy/*.py
pycodestyle ./hsimpy/*.py
```




### TESTS

After package installing, into folder of project enter below command.

``` bash
virtualenv hsimp_env
pip3 install -r requirements-test.txt
python3 -m unittest discover tests  # will discover al tests

```




## REFERENCES

- https://howsecureismypassword.net/
- https://github.com/howsecureismypassword/



## AUTOR

> [contribuidores](./CONTRIBUTING.md)


CHANGELOG.md
|    date    |  version  |              description             |
|------------|-----------|--------------------------------------|
| 16.19.2019 |   0.0.5   | [+] starting support py3             |
| 31.01.2020 |   0.0.6   | [*] update to alg force              |
| 19.03.2020 |   1.0.0   | [*] v1 to python3                    |


tests/test_hsimpy.py
# -*- coding: utf-8 -*-


import unittest
from hsimpy import Hsimpy


class HsimpyTest(unittest.TestCase):
    '''
    TESTING TYPE OF password PARAM
    '''
    def test_password_function_param(self):
        try:
            def func():
                return

            hsimp_func = Hsimpy(func)
        except TypeError as err:
            self.assertRaises(TypeError)

    def test_password_int_param(self):
        try:
            hsimp_number = Hsimpy(123456);
        except TypeError as err:
            self.assertRaises(TypeError)

    def test_password_float_param(self):
        try:
            hsimp_float = Hsimpy(123.456)
        except TypeError as err:
            self.assertRaises(TypeError)

    def test_password_dict_param(self):
        try:
            hsimp_dict = Hsimpy({dict: 2})
        except TypeError as err:
            self.assertRaises(TypeError)

    def test_password_json_param(self):
        try:
            hsimp_json = Hsimpy({"dict": "2"})
        except TypeError as err:
            self.assertRaises(TypeError)

    def test_password_unicode_param(self):
        hsimp_unicode = Hsimpy(u'dale')
        self.assertEqual(True, True)

    def test_password_str_param(self):
        hsimp_str = Hsimpy('dale')
        self.assertEqual(True, True)

    '''
    TESTING STRENGHT algorithm
    '''
    def test_password_bad_str_param(self):
        hsimp = Hsimpy('weak')
        self.assertEqual(hsimp.security_level, 'bad')
        self.assertEqual(hsimp.possible_characters, 26)
        self.assertEqual(hsimp.possible_combinations, 456976.0)
        self.assertEqual(hsimp.time_in_seconds, 0.11717333333333334)

    def test_password_bad_str_up_param(self):
        hsimp = Hsimpy('weakWEAK')
        self.assertEqual(hsimp.security_level, 'bad')
        self.assertEqual(hsimp.possible_characters, 52)
        self.assertEqual(hsimp.possible_combinations, 53459728531456.0)
        self.assertEqual(hsimp.time_in_seconds, 13707622.700373333)

    def test_password_ok_str_up_number_param(self):
        hsimp = Hsimpy('weakWEAK123')
        self.assertEqual(hsimp.security_level, 'good')
        self.assertEqual(hsimp.possible_characters, 62)
        self.assertEqual(hsimp.possible_combinations, 5.20365606838371e+19)
        self.assertEqual(hsimp.time_in_seconds, 13342707867650.537)

    def test_password_good_str_up_number_special_param(self):
        hsimp = Hsimpy('weakWEAK123!@#')
        self.assertEqual(hsimp.security_level, 'good')
        self.assertEqual(hsimp.possible_characters, 77)
        self.assertEqual(hsimp.possible_combinations, 2.575550990467221e+26)
        self.assertEqual(hsimp.time_in_seconds, 6.6039768986339e+19)

    def test_password_good_str_up_number_special_dots_param(self):
        hsimp = Hsimpy('weakWEAK123!@#.,')
        self.assertEqual(hsimp.security_level, 'good')
        self.assertEqual(hsimp.possible_characters, 96)
        self.assertEqual(hsimp.possible_combinations, 5.204029246664727e+31)
        self.assertEqual(hsimp.time_in_seconds, 1.334366473503776e+25)

    def test_password_bad_number_param(self):
        hsimp = Hsimpy('123')
        self.assertEqual(hsimp.security_level, 'bad')
        self.assertEqual(hsimp.possible_characters, 10)
        self.assertEqual(hsimp.possible_combinations, 1000.0)
        self.assertEqual(hsimp.time_in_seconds, 0.0002564102564102564)

    def test_password_bad_number_special_param(self):
        hsimp = Hsimpy('123!@#')
        self.assertEqual(hsimp.security_level, 'bad')
        self.assertEqual(hsimp.possible_characters, 25)
        self.assertEqual(hsimp.possible_combinations, 244140625.0)
        self.assertEqual(hsimp.time_in_seconds, 62.600160256410255)

    def test_password_bad_number_special_dot_param(self):
        hsimp = Hsimpy('123!@#.,')
        self.assertEqual(hsimp.security_level, 'bad')
        self.assertEqual(hsimp.possible_characters, 44)
        self.assertEqual(hsimp.possible_combinations, 14048223625216.0)
        self.assertEqual(hsimp.time_in_seconds, 3602108.6218502563)

    def test_password_ok_number_special_dot_upper_param(self):
        hsimp = Hsimpy('123!@#.,ABC')
        self.assertEqual(hsimp.security_level, 'good')
        self.assertEqual(hsimp.possible_characters, 70)
        self.assertEqual(hsimp.possible_combinations, 1.977326743e+20)
        self.assertEqual(hsimp.time_in_seconds, 50700685717948.72)

    def test_password_bad_special_param(self):
        hsimp = Hsimpy('!@#$')
        self.assertEqual(hsimp.security_level, 'bad')
        self.assertEqual(hsimp.possible_characters, 15)
        self.assertEqual(hsimp.possible_combinations, 50625.0)
        self.assertEqual(hsimp.time_in_seconds, 0.012980769230769231)

    def test_password_bad_dot_param(self):
        hsimp = Hsimpy('.,.')
        self.assertEqual(hsimp.security_level, 'bad')
        self.assertEqual(hsimp.possible_characters, 19)
        self.assertEqual(hsimp.possible_combinations, 6859.0)
        self.assertEqual(hsimp.time_in_seconds, 0.0017587179487179487)

    def test_password_ok_str_repeat_param(self):
        hsimp = Hsimpy('aaaaaaaaaaaaaa')
        self.assertEqual(hsimp.security_level, 'good')
        self.assertEqual(hsimp.possible_characters, 26)
        self.assertEqual(hsimp.possible_combinations, 6.450997470329715e+19)
        self.assertEqual(hsimp.time_in_seconds, 16541019154691.578)

    def test_password_ok_number_repeat_param(self):
        hsimp = Hsimpy('1111111111111111111')
        self.assertEqual(hsimp.security_level, 'good')
        self.assertEqual(hsimp.possible_characters, 10)
        self.assertEqual(hsimp.possible_combinations, 1e+19)
        self.assertEqual(hsimp.time_in_seconds, 2564102564102.564)



if __name__ == '__main__':
    unittest.main()


hsimpy/hsimpy.py
# -*- coding: utf-8 -*-


'''
How Secure is My Password? Python Module
'''


import math
import re


class Character():
    '''
    Character module will be help to know the possibility combinations

    Parameters
    ----------
    password : str
        the string contains the password value.

    Attributes
    ----------
    contained_in_sets : list
        List of char type is been used
    possible_characters : int/float
        Number os possible characters combination.
    character_sets_dictionary : dict
        Configuration of test char
    '''

    def __init__(self, password):
        if not isinstance(password, str):
            raise Exception('password is an invalid type', password)

        self.contained_in_sets = []
        self.possible_characters = 0
        self.character_sets_dictionary = {
            "ASCII Control Character": ["[\\u0000-\\u001F]", 32],
            "ASCII Lowercase": ["[a-z]", 26],
            "ASCII Uppercase": ["[A-Z]", 26],
            "ASCII Numbers": ["\\d", 10],
            "ASCII Top Row Symbols": ["[-!@£#$%^&*()=+_]", 15],
            "ASCII Other Symbols": ["[\\s\\?\\/\\.>,<`~\\|;:\\]}\\[{'\"\\\\]", 19],
            "Unicode Latin 1 Supplement": ["[\\u00A1-\\u00A2\\u00A4-\\u00FF]", 93],
            "Unicode Latin 1 Supplement Non Standard": ["[\\u0080-\\u00A0]", 33],
            "Unicode Latin Extended A": ["[\\u0100-\\u017F]", 128],
            "Unicode Latin Extended B": ["[\\u0180-\\u024F]", 208],
            "Unicode Latin Extended C": ["[\\u2C60-\\u2C7F]", 32],
            "Unicode Latin Extended D": ["[\\uA720-\\uA7FF]", 29],
            "Unicode Cyrillic Uppercase": ["[\\u0410-\\u042F]", 32],
            "Unicode Cyrillic Lowercase": ["[\\u0430-\\u044F]", 32]
        }

        for item in self.character_sets_dictionary.items():
            key = item[0]
            value = self.character_sets_dictionary[key]
            re_value = value[0]
            possibility = value[1]

            length = len(password)
            password = re.sub(r'%s' % re_value, '', password)

            if not length == len(password):
                self.contained_in_sets.append(key)
                self.possible_characters += possibility

        if len(password) > 0:
            self.contained_in_sets.append('Unspecified Unicode')
            self.possible_characters += 1e3


class Hsimpy():
    '''
    Rather than just saying a password is "weak" or "strong",
    *How Secure is My Password?* lets your users know how long it would
    take someone to crack their password.

    Parameters
    ----------
    password : str
        the string contains the password value.

    Attributes
    ----------
    conf : dict
        COnfiguration using to calc the strength
    possible_characters : int
        Number os possible character.
    possible_combinations : int/float
        Number os possible characters combination.
    time_in_seconds : int/float
        Time to broke passwork using 10bilions operation p/s
    security_level : str
        bad, ok, good
    '''

    def __init__(self, password):
        # import pdb; pdb.set_trace()
        if not isinstance(password, str):
            raise TypeError('password is an invalid type', password)

        self.conf = {
            "calculations_per_second": 3900000,  # 1e10 10 billion
            "good": 31557600e3,  # 31557600e6 1 million years
            "ok": 31557600e1  # 31557600 1 year
        }

        char = Character(password)

        self.possible_characters = char.possible_characters
        self.possible_combinations = math.pow(self.possible_characters, len(password))
        self.time_in_seconds = self.possible_combinations / self.conf['calculations_per_second']
        self.security_level = 'bad'

        if self.time_in_seconds >= self.conf['good']:
            self.security_level = 'good'
        elif self.time_in_seconds >= self.conf['ok']:
            self.security_level = 'ok'
